### ISSUE TITLE
Use config for color module palettes

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#2554]
Color module uses an array in a color.inc file to store palettes. If we used config, people could import new palettes to their themes, have multiple custom palettes, and easily export color settings from dev to prod.

Could this work before 2.0? It would be an API change?

**Comment by klonos** [comment]
This makes so much sense! We need to support having `/config` directories in themes (does this already work BTW?). ...but would that then also require themes to have install/uninstall hooks (so that the config files are removed when removing a theme)? Which might mean that we'd need to introduce an "Uninstall themes" tab in the UI.

**Comment by herbdool** [comment]
Themes already store some things in config files (which you can't actually uninstall) so the basics are there.

Maybe you could just build the array by calling config->get

**Comment by docwilmot** [comment]
@herbdool its not the technical stuff I'm concerned about but the API change for contrib. Although, practically speaking, all contrib with color support were made by me 😉.

**Comment by herbdool** [comment]
Good idea @hosef

**Comment by klonos** [comment]
Yep, makes sense @hosef 👍

**Comment by docwilmot** [comment]
Proof of concept. 
- Saves a JSON version of a theme's `color.inc` in config when the theme is installed
- Allows saving custom color palettes.

This doesnt yet:
- Allow themes to ship with a config file to sync configs from themes to active. Couldnt decide if we should have a theme `config` folder, or if we should scan the `color` folder for configs, or if a config folder within the `color` folder, etc.
- Allow to edit or delete saved custom color palettes via UI.

**Comment by docwilmot** [comment]
The code for this was updated a while back, but needs some explanation. With this PR, `color_get_info()` has been updated to look for color info both in config and in `color.inc`:
- existing themes using `color.inc` files would work normally, color being pulled from `color.inc`. Once the themes setting is saved a new `themeNAME.color.default.json` file is created in active config. Backdrop looks for this file first, and then checks for a `color.inc`, so future reads for color info will be from this config file.
- new themes can ship with a `themeNAME.color.default.json` file, and this will be synced to config and be the source of all color info whether there is a `color.inc` or not.
- themes or module contrib can ship with additional `themeNAME.color.STRING.json` files that can *add* new color presets (detail later). These will be merged by `color_get_info()`, so once theme settings are saved, all theme color data from all configs will end up saved in `themeNAME.color.default.json`. There may be some debate as to whether we should keep those separate.
- what would have been really nice (planned in my mind all along) would be if contrib modules could provide color scheme presets for their own CSS files. Say you build a new slider but would like for it to work well and look good with all Basis color presets (+/- provide your own). But alas this may not work, as color matching is only done on a CSS file that is declared in the ``themeNAME.info` file! Would perhaps be helpful to provide a hook that adds contrib CSS files to the ones that Color module parses?

Re additional color files, don't know how helpful this would be but it was just 3 additional lines to enable the functionality so it's in there. So if you found a nice color scheme you wanted to share, you can create a new color config file. For example (and this probably qualifies as a bug) our new front page cards have a `cards.css` file, but its not listed in Basis `color.inc` `css` array. The following config file now allows Backdrop to dynamically change the card background. It declares a new color field (cardbackground), sets a default (the base color that Color module will replace) and one scheme (the scheme will automatically show up in Basis settings and if if you switch Basis to that scheme, card Backgrounds change to `#006600`. Note we use this config to add `cards.css` to the list of CSS file targets.
```
{
    "_config_name": "basis.color.cards",
    "fields": {
        "cardbackground": "Card background"
    },
    "schemes": {
        "default": {
            "colors": {
                "cardbackground": "#f2f2f2"
            }
        },
        "card_scheme": {
            "title": "Card scheme",
            "colors": {
                "cardbackground": "#006600"
            }
        }
    },
    "css": [
        "cards.css"
    ]
}
```

And that's all. Although probably not a very used feature, but update color to new standards.

**Comment by docwilmot** [comment]
Could I get a milestone and/or review on this please? Tests pass so far.

**Comment by klonos** [comment]
No idea why this was assigned to me 🤷🏼

The issue already has the `milestone candidate - minor` label. As per our [documentation](https://docs.backdropcms.org/documentation/adding-milestones-to-issues#minor-milestones), in order to get the actual milestone set for a minor version, we also need an advocate. @docwilmot (or others) if you are advocating for this feature, then please edit the issue summary to indicate that, and we can then take it from there. 

Also, does this depend on #6122 in any way? (or the other way around?)

**Comment by klonos** [comment]
OK, I have some questions (but I might simply need to spend more time testing this in order to get a better understanding).

1. Why do we need to keep `color.inc`? If we provide an upgrade path from that to using only config instead, then couldn't we ditch it? Having multiple ways to do the same thing can become confusing, it complicates code, as well as our documentation. I was thinking that this could work along these lines:
  - if we detect a `color.inc` file provided by a theme, we convert it to config, then we ignore that `color.inc` file + throw a warning in the status page and/or the theme settings page.
  - in Backdrop 2.x, we can consider removing the code that provides the functionality to convert `color.inc` files - the expectation will be that all themes would have switched to config by then.
2. In the current implementation of the feature in the PR, if the same color scheme happens to be defined in `color.inc` + the default config file + any additional config file(s), which one gets precedence/priority? Which one should be the source of truth, and how can a themer know that?

@docwilmot can you please outline various scenarios to test this? What would be a simple use case? Anything more complex?

**Comment by hosef** [comment]
1. We should probably keep it for backwards compatibility. Currently, if you add a new color to an existing theme through color.inc it will show up in the UI right away. That should continue to happen for themes with a color.inc.

**Comment by hosef** [comment]
I have been looking through this for a bit, and I think that the color.inc file should continue to be the source of truth if it exists. In the PR, if it finds a config file it will ignore the color.inc file in favor of using the config file. I can easily imagine a scenario where someone might spend hours trying to find what is going on.

* Developer creates a color.inc file and starts adding settings
* Developer sees that as settings are added, the UI updates automatically
* Developer presses the save button 1 time to test if it is working
* Developer spends hours trying to figure out why changes to color.inc are no longer showing up in the UI
*

**Comment by docwilmot** [comment]
> we also need an advocate

I've advocated.
> Also, does this depend on https://github.com/backdrop/backdrop-issues/issues/6122 in any way?

Note #6122 is already merged, its only open for docs. And yes this would depend on that.
> Why do we need to keep color.inc? 

Because `color.inc` is provided by existing themes, not core, and at the time this PR this is merged, no existing theme will have its color info in `config` yet. And maybe even for years to come few theme devs will bother to use config either. 
> if we detect a color.inc file provided by a theme, we convert it to config, then we ignore that color.inc file + throw a warning in the status page and/or the theme settings page.

This is essentially what this PR does: check for color info in either `color.inc` or config, ignore `color.inc` if a config is found. This way theme devs can either convert their themes to config, and delete the `color.info` file and create a new release, or if theyre nervous about this new-fangled system, leave both in, let Backdrop sort out which is used. No warning message, then, since its not a mistake, and requiring end-users to"hack core" and delete the `color.inc` file manually to get rid of a persistent warning would be excessive.
> I think that the color.inc file should continue to be the source of truth if it exists.

I think this is contrary to what we do with other theme settings though: we have a hard-coded `settings[]=''` in theme info files, but once you've saved your theme settings, Backdrop uses the config file thereafter. I think this is the reasonable way to go because the opposite defeats the purpose of portability of config.
> Developer spends hours trying to figure out why changes to color.inc are no longer showing up in the UI

Perhaps  re-merge the two on cache flush? Or maybe add a message "Color info is being read from config" on settings save might be simpler.

**Comment by klonos** [comment]
> ... No warning message, then, since its not a mistake, ...Or maybe add a message "Color info is being read from config" on settings save might be simpler.

Yes, I like that 👍🏼 ...same as we are throwing a message to let people know that we have flushed caches when they switch themes.

However, since this is not a one-time thing like cache clear, the message should be in the settings page of the theme constantly instead of shown only once during save. This should be more like the info message we are showing when the Color module is disabled and a theme supports it:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/cb69a3e0-3cd0-4d98-b7b3-2d8c0ae1bf15)

**Comment by hosef** [comment]
I guess my main question is, how do theme developers update this after the theme settings are saved? Because the settings need to reflect what is in the CSS files provided by the theme, then any change to the CSS would require changes to the settings. Suppose the theme developers want to add a new setting, or change the original color in CSS, or change a preset that they provided.

My guess is that we have 2 options. The first is that we would need to allow theme developers to put update hooks in the theme. I suspect that would almost guarantee that no theme developers would actually use the system. The second would be to make a UI for the color module settings. If we are going to go through the effort to do that though, I would recommend making something more comprehensive in contrib instead.

**Comment by docwilmot** [comment]
Not quite following you @hosef. Are you meaning for a theme out in use and the dev has made updates in a new release? As I understand it that dev would have needed to update the `color.inc` in that case, and would now instead update the json file instead?

**Comment by docwilmot** [comment]
I think this is ready for another review please.

**Comment by argiepiano** [comment]
@docwilmot, can you outline ways to test this?

**Comment by argiepiano** [comment]
Some prelim testing. I created a new custom palette with a name.
- I see the new color config file in the config folder called THEME.color.default.json, which includes new palette 👍🏽 
- It's strange that new palettes are included in the `color.default` file. I had assumed that it would create separate config files for each new palette, making it easier to import/export
- More serious: I don't see the new color config file listed in admin/config/development/configuration/single/export
- And, if I uninstall the theme, the THEME.color.default.json config file is not removed

**Comment by docwilmot** [comment]
Thanks for testing, I believe all above should be addressed now.
> @docwilmot, can you outline ways to test this?

In a theme that is color-enabled, when you go to the color settings form:
- When you select "Custom" color scheme a name field should pop up, which will allow you to save your new color scheme
- Same happens if you modify a color on a theme-provided color scheme
- New custom schemes will be editable
- Custom schemes will save to separate `THEME.color.SCHEME_NAME` files in config
- Enabling a theme should create a `THEME.color.default` file in config
- For existing installs, there is an upgrade hook to create `THEME.color.default` file in config. But just saving the appearance page should do it as well.
- Disabling a theme should delete all color configs
- All schemes should be exportable

**Comment by docwilmot** [comment]
I'm thinking in the schemes list custom schemes could use a suffix to show theyre custom? `my-scheme - (custom)` for example. Not sure.

**Comment by argiepiano** [comment]
More testing here. This is working great! Thanks for fixing my above comments.

Additional ones:  

1. When I select Custom, enter the name of the new scheme, and click save **without changing any color,** the new scheme is not selected. I have to use the select input widget once again to choose the newly created scheme. Is there any way to select the newly created scheme in the submit handler?  (BTW, if you do change a color in the new scheme before saving, it does get selected)
2. This is more of a hypothetical question: I noticed all the json files generated for the schemes contain a lot of the same elements, which do not change regardless of the changes you make to the colors. Those are: `css`, `fields`, and `blend_target`. Wondering if it's necessary to save this same info every time? Seems redundant. I realize this may mean splitting the default file as well, perhaps into two json files: `basis.color.settings.json` and `basis.color.SCHEME.json` (and SCHEME may be `default` as well). 
3. 
4. Also, I realize that there is another json file: `basis.settings.json` which contains a different set of elements (this already existed and is separate from this issue). But perhaps instead of my suggested `basis.color.settings.json` we could store `fields`, `css` and `blend_target` in that json file, since these do not change with each scheme?

**Comment by indigoxela** [comment]
@docwilmot what does this change mean to contrib (or custom) themes with color module support (and color schemes, currently in code)?

They all need to get updated, am I getting this right?

And given, that themes can't have update hooks - what would happen to switch those from php code to json config?

**Comment by docwilmot** [comment]
> When I...click save without changing any color, the new scheme is not selected

Color decides which scheme is active by checking the current palette. If you simply saved an existing scheme under a new name, Color might calculate that the old scheme is still active because its exact palette is active. I'll try and figure how to get around that.
>  But perhaps...we could store fields, css and blend_target in that json file, since these do not change with each scheme

This makes sense, will look into that
> They all need to get updated, am I getting this right?

No they wont; Color module will still read from an existing `color.inc` file as normal.
> And as themes can't have update hooks - what would have to happen to switch those from php code to json config?

I've thought about this, and @hosef questioned similar. 
- If a theme uses config (with a config folder and `x.color.x.json` files), then new installs will just run normally of course with the new system. 
- If an existing install updates to the new version of a currently installed theme where this new version has a new config folder (site owner just swaps out the old theme directory), nothing new will happen since config is only installed to `active` on enabling a new theme, and the theme will continue to read from `color.inc`. But I think "nothing happens" is preferable with themes though, even without this PR. You dont want your site suddenly looking different unless you want to. In this case, if you want to, you disable and re-enable the theme to get new config, or import config, or copy the config files to your `active` folder (or whatever process you use).
- In the future, when a site is already using config, and updates to the latest version of the theme (with new `color` configs in the `config` folder) I suspect you'd need to import the new configs manually as well, unless we build in some method of checking if there are unimported configs and prompt the user to use import.
- Maybe other scenarios

**Comment by docwilmot** [comment]
Updated PR:
> When I...click save without changing any color, the new scheme is not selected

This should be fixed.
> But perhaps...we could store fields, css and blend_target in that json file, since these do not change with each scheme

Now saving all color info in the existing theme settings file `color` array (which seems more sensible). Custom schemes now only have a `schemes` array (and a `scheme_config_name` which is there for `hook_config_info()` to find these files). Custom color config files can store more than one scheme.

**Comment by stpaultim** [comment]
Added this to 1.28 milestone, since it has an advocate.

**Comment by stpaultim** [comment]
It seems a bit strange that it counts existing schemes before starting to number custom schemes. The first custom scheme is numbered as 7, because of 6 other provided schemes. 

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/72b24586-5298-4ebd-a90b-1681e8badc05)

**Comment by stpaultim** [comment]
It took me several times of experimenting with this PR to realize that every time I edit one of the default color schemes, I am given the opportunity to rename it. I ONLY noticed this option, when selecting the custom option. I think this was purely about my own expectations. But, we might want to do something to make the option to rename custom schemes more visible or even to require it. 

The first time I edit a default color scheme without giving it a new name, the PR creates a custom scheme with the label "Custom 7". But, when I go back and edit that custom scheme, the label changes from "Custom 7" to "custom_7". That seems like a bug. 

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/a91263c7-72f1-4bc5-bae1-eb24bc1229bb)

**Comment by kiamlaluno** [comment]
> It took me several times of experimenting with this PR to realize that every time I edit one of the default color schemes, I am given the opportunity to rename it.

It is only possible to set a name when selecting *Custom* as color scheme. In that case, the field for the scheme theme is right after the scheme selector.

![202403-260858](https://github.com/backdrop/backdrop-issues/assets/29851/161025c2-c4a7-43e0-9383-4d305ceb399b)

For the other color schemes, I cannot see any field to enter a scheme name.

I would like to create a new color scheme from, for example, *Plume*, but I would suggest that in a different issue.

**Comment by kiamlaluno** [comment]
As for the schema name change, I noticed two changes: After I click on *Save theme settings*, the scheme selector shows *red skin* instead of *Red Skin* but the list of schemes shows *red_skin*.

**Comment by docwilmot** [comment]
> This is what happens when I try to name my custom scheme "Tim's Theme".

Fixed
> But if you edit an existing scheme it get's an automatic name that can't be changed in the UI.
> I had a custom theme named "custom_7" and "Custom 8"
> But, when I go back and edit that custom scheme, the label changes from "Custom 7" to "custom_7". That seems like a bug
> After I click on Save theme settings, the scheme selector shows red skin instead of Red Skin but the list of schemes shows red_skin.

Fixed I believe.
> It seems a bit strange that it counts existing schemes before starting to number custom schemes. The first custom scheme is numbered as 7, because of 6 other provided schemes.

Fixed. Starts numbering based on custom theme count only.
> It took me several times of experimenting with this PR to realize that every time I edit one of the default color schemes, I am given the opportunity to rename it...But, we might want to do something to make the option to rename custom schemes more visible or even to require it.
> It is only possible to set a name when selecting Custom as color scheme. In that case, the field for the scheme name is right after the scheme selector.

So you can either select "Custom" first and you get the field to add a name. Or once you change a color on an existing default theme the field also pops up and you can name this as a new scheme. But if you change a colo on and existing *custom* scheme, thats considered editing a custom scheme and you shouldnt be able to rename it then.

**Deleting and renaming**
So far the UI doesnt allow to rename custom schemes nor delete them. This could easily be a followup, but it shouldnt be hard. The UI to do so would be the decision though.

**Comment by stpaultim** [comment]
Did I miss the answer to @indigoxela 's question about how this effects contrib themes?
https://github.com/backdrop/backdrop-issues/issues/2554#issuecomment-1996660918

**Comment by docwilmot** [comment]
> What if new custom schemes were named after the original scheme that was modified. If I modify the "Goldfish" scheme, the new scheme could be "Goldfish-1" by default with the option to give it a comletely new name. The advantage of this would be in how schemes would be organized in the dropdown.

This is a great idea, but it would require a JS solution I think, which may take more time. 

> Did I miss the answer to @indigoxela 's question about how this effects contrib themes?

I answered in https://github.com/backdrop/backdrop-issues/issues/2554#issuecomment-1997479653 I believe.

**Comment by indigoxela** [comment]
> I answered in https://github.com/backdrop/backdrop-issues/issues/2554#issuecomment-1997479653 I believe.

Yes, I think it has been answered. However, this needs actual testing, which I can't do currently, as I'm involved in getting several other issues done for 1.28 (testing / review...).
As a maintainer of several contrib themes with color support, I wouldn't want to update all of them (no time for that now), or - worst case - that something breaks for those themes or their subthemes. If that's not the case, I'm fine.

**Comment by quicksketch** [comment]
This is looking pretty good!

@indigoxela I don't think any adjustment is needed by contrib themes. From the PR, there are no changes to Basis, and yet it gains the ability to store color palettes in config. `color.inc` should be deprecated though, and we can remove it in 2.x.

@docwilmot An item I found is that color module should clean up a theme's color settings through `hook_themes_disabled()`. Same as system module cleans up `[theme].settings` config items on disable in [system_themes_disabled()](https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.module/function/system_themes_disabled/1).

**Comment by quicksketch** [comment]
Initial local testing of this PR, I found that it seems to break the "Blue Lagoon" color scheme preview.

* On an existing site, pull down this PR, clear caches.
* Enable the Bartik theme.
* Visit `admin/appearance/settings/bartik` and change the scheme to "Blue Lagoon" and click "Save".
* The page reloads, the Blue Lagoon style _does_ get applied to the front-end of the site, but the preview is now not showing any colors:

![image](https://github.com/backdrop/backdrop-issues/assets/100206/f735dce9-e02c-4963-b1fd-cf1d1acc19f2)

Looking in my active config, this is the `bartik.settings` config file:
```
{
    "_config_name": "bartik.settings",
    "main_menu_tabs": "rounded-tabs",
    "custom_name": "",
    "color_legacy": true
}
```

If I change the color scheme to something else, like "Mocha", then it starts working again. Set it back to "Blue Lagoon" and the preview breaks again.

**Comment by quicksketch** [comment]
Looking at the exported configuration... I guess I was expecting to just find the color schemes specified in the config file, not every aspect of the `color.inc` specification in the config export. It seems like a lot of stuff in there isn't really "config", since it's fundamentally tied to the file structure and CSS of the theme. Although if we didn't export everything out of `config.inc`, then we couldn't deprecate the file. How would you feel about just pulling and merging the contents of the `schemes` array from config rather than everything? I might also not copy out the core schemes and have this just be additive rather than replacing all the schemes provided by color.inc.

**Comment by stpaultim** [comment]
I finally did a test on a local environment, so I could see the config files. I'm not sure I understand the comments by @quicksketch 

In my first test, it creates a config file that looks like this. I think I modified the Plum scheme in Basis.

```
{
    "_config_name": "basis.color.tim_s_scheme",
    "schemes": {
        "tim_s_scheme": {
            "title": "Tim's Scheme",
            "colors": {
                "header": "#4c1c58",
                "base": "#000100",
                "slogan": "#9512ba",
                "titleslogan": "#ffffff",
                "hovermenu": "#9d408e",
                "menutoggle": "#2c2c28",
                "bg": "#fffdf7",
                "text": "#471515",
                "link": "#9d408d",
                "borders": "#2c2c28",
                "formfocusborder": "#edede7",
                "primarytabs": "#2b2b2b",
                "primarytabstext": "#301313",
                "buttons": "#c9d9e3",
                "footerborder": "#89297a",
                "footer": "#cca8a8",
                "footertext": "#000100"
            }
        }
    },
    "scheme_config_name": "tim_s_scheme"
}
```

**Comment by quicksketch** [comment]
Responding to my own question:

> How would you feel about just pulling and merging the contents of the schemes array from config rather than everything?

Ah, I now see what's happening here. The new color schemes are saved to separate config files, i.e. `basis.color.[name].json`, I hadn't noticed that initially. I was only looking at the changes in the theme overall settings config, i.e. `basis.settings.json`.

So my suggestion of being "additive" with additional palettes is already in place. But I'm still not sure about replacing `color.inc` with settings in config.

One area that could be particularly problematic is that themes don't have the ability to run update hooks, so if the config format were ever to change (say the theme added a new color to the palette), it wouldn't have any way to update the active config. A theme also wouldn't have the ability to expand to include new palettes. After the theme is installed, it would never be able to make any changes.

**Comment by docwilmot** [comment]
@quicksketch, yes, the effect of this PR is to save all the info in `color.inc` to the theme's settings file. Then custom schemes (only the color scheme info as in @stpaultim comment above) get saved to separate config files. How this evolved was first I considered as you mentioned, splitting just the schemes into config, but the defaults shouldnt be editable I thought, so we'd end up with a `basis.color.defaults` with all the defaults, and then separate `basis.color.customname` files, which seems like it would be a bit cumbersome for themes to have two settings files (for settings and separate for color info).
> But I'm still not sure about replacing color.inc with settings in config.

Maybe instead a `hook_color.info()` in `template.php`?
> One area that could be particularly problematic is that themes don't have the ability to run update hooks, so if the config format were ever to change (say the theme added a new color to the palette), it wouldn't have any way to update the active config

We discussed this somewhere earlier; my thinking is that if a theme wanted to add new palettes they would need to be imported. Or we could make it so that saving theme settings pulls any new config files into active?

Either case I suspect will need to rethink this for few more months. I dont think Color is that popular, so I dont feel very driven to push this. Will try again next feature freeze.

**Comment by quicksketch** [comment]
> Maybe instead a `hook_color_info()` in template.php?

I generally like this kind of direction. I don't like the `color.inc` or `theme-settings.php` files, but I'm not sure what we really gain much rearranging the code. There would need to be some benefit to it as well.

**Comment by kiamlaluno** [comment]
Using a hook would allow to also use an alter hook; the information could also be dynamically built. (I guess those are the usual reasons for using a hook instead of reading from a data file.)

**Comment by stpaultim** [comment]
I'd be interested in a clear update on the status of this issue.

What is blocking it?
Is there still a sense that this is a good candidate for core it something like its current state?

**Comment by docwilmot** [comment]
> What is blocking it?

This needs a re-think. If I recall correctly the main issue is that we are currently trying to store data in config that is probably best in code ideally in a hook implementation. We will probably be able to get it sorted for the next release.

**Comment by jenlampton** [comment]
I'm going to pull this from the 1.30 milestone since it's not likely to get in, but after the re-think feel free to add it back to the next one :)  
It might be worth mentioning that the color module really needs to be rewritten to use CSS variables, and if we do that, it could spark a handful of other questions that may need to be answered.
